designs/CentrallyManagingNuGetPackageVersions-VSManagePackages.md

```diff
@@ -0,0 +1,95 @@
+
+There are two categories of solutions.
+
+* API ignorant of centrally managed package versions
+
+Project system APIs would not be changed, but the implementation would need to be aware of how to detect if the feature is enabled, and know when to add a version to the `PackageReference` and when not to, and when to modify the `Directory.Packages.props` and when not not.
+
+There would also be a potential issue when a single user gesture causes a package to be installed into multiple projects, particularly when the `Directory.Packages.props` file does not already contain a `PackageVersion` for the package. If NuGet calls APIs to modify all the projects in parallel, there might be a race condition with trying to update `Directory.Packages.props`.
+
+* API aware of centrally managed package versions
+
+Presumably this would mean there's a separate API to add a `PackageVersion` to the `Directory.Packages.props` file. But since this file is shared across multiple projects, would the API be independent from the projects, or should NuGet pick one project and call the API on that project?
+
+Should new APIs be added to the project systems considering the version attribute must not be written? Or NuGet could call the API with a null version. Or NuGet passes the version and the project system must know if the project uses centrally managed package versions or not.
+
+### Removing a package from a project
+
+Currently NuGet calls APIs on the project system to remove the `PackageReference` for that package given a package id. The current API is sufficient and the same behavior should continue. The `Directory.Packages.props` file should not be modified to remove the `PackageVersion`, as other projects may still be using it.
+
+### Change the version of a package in `Directory.Packages.props`
+
+This will affect, at minimum, all the loaded projects in the solution that have a `PackageReference` to the package whose version is changed. For SDK-style projects, it appears this should work already, by watching the file for changes and then sending NuGet a project nomination with the new information. Currently non-SDK style projects need to be unloaded and reloaded to take into account changes to imported MSBuild files.
+
+Given this single user gesture will often affect multiple projects, should the API to modify the `Directory.Packages.props` file be on a project system interface, or a new interface that's decoupled from individual projects?
+
+Once the file is modified, should NuGet notify projects that it's changed, or should they detect it themselves?
+
+In particular, given that SDK style projects currently use file watchers to detect changes to imported files and automatically reload, but non-SDK style projects do not and continue to use outdated information until the project is reloaded, how can we avoid a bad user experience when a customer uses Package Manager UI/Console to change package versions?
```

Comment:
Currently if a project file is updated outside visual Studio with a new package Version , Visual Studio detects the change and asks for reload.
We can go on the same route or only present the buttons "Reload All" / "Ignore All"  .

Reply:
The non-SDK project system prompts to reload the project only if the project file is updated. If a file that is imported by the project file is modified, the project system ignores the change until the project is reloaded. This is a problem for the `Directory.Packages.props` file.

If project system wants to start watching imported files for changes, great. Forcing customers to reload the project every time any project in the solution updates the props file will give a bad experience, but at least it won't restore the wrong packages. If NuGet is responsible for monitoring the files for changes, and then forcing the project system to reload, well, if we have to, then it's probably better than restoring the wrong version of packages.  But this is something we need to talk to project system teams about. Does anyone have any ideas how to give a better customer experience? If changes to APIs are needed, will anyone commit to it?

Reply:
We'd still need to account for the scenario where we might kick off a new restore before all nominations come in. That could lead to a restore failure and then a successful restore later when all the nominations arrive. 

This becomes more complex with the csproj.dll based projects.